File: routes/stats.js
const express = require('express');
const router = express.Router();
const Todo = require('../models/Todo');

// ─── GET /api/stats ───────────────────────────────────────────────────────────
router.get('/', async (req, res) => {
  try {
    const [total, completed, overdue, byPriority] = await Promise.all([
      Todo.countDocuments(),
      Todo.countDocuments({ completed: true }),
      Todo.countDocuments({ completed: false, dueDate: { $lt: new Date() } }),
      Todo.aggregate([
        { $group: { _id: '$priority', count: { $sum: 1 } } },
      ]),
    ]);

    const pending = total - completed;
    const completionRate = total > 0 ? Math.round((completed / total) * 100) : 0;

    const priorityMap = { low: 0, medium: 0, high: 0, urgent: 0 };
    byPriority.forEach(({ _id, count }) => {
      if (_id in priorityMap) priorityMap[_id] = count;
    });

    res.json({
      success: true,
      data: {
        total,
        completed,
        pending,
        overdue,
        completionRate,
        byPriority: priorityMap,
      },
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
});

module.exports = router;

File: routers/stats.py
"""
routers/stats.py
────────────────
Aggregated stats used to power the sidebar dashboard.

Endpoint:
    GET /api/stats  →  { total, completed, pending, overdue, completion_rate, by_priority }
"""

from datetime import datetime, timezone
from fastapi import APIRouter
from models.todo import Todo, Priority

router = APIRouter()


@router.get("/")
async def get_stats():
    now = datetime.now(timezone.utc)

    total     = await Todo.count()
    completed = await Todo.find(Todo.completed == True).count()
    pending   = total - completed
    overdue   = await Todo.find(
        Todo.completed == False,
        Todo.due_date < now,      # type: ignore
    ).count()

    completion_rate = round((completed / total) * 100) if total > 0 else 0

    by_priority = {}
    for p in Priority:
        by_priority[p.value] = await Todo.find(Todo.priority == p).count()

    return {
        "success": True,
        "data": {
            "total":           total,
            "completed":       completed,
            "pending":         pending,
            "overdue":         overdue,
            "completion_rate": completion_rate,
            "by_priority":     by_priority,
        },
    }
